File: Problems/1372. (M) Longest ZigZag Path in Binary Tree/Java/solution.java
class Solution {
  // recursive DFS

  int pathLength = 0;

  public int longestZigZag(TreeNode root) {
    // root has no parent, so direction does not matter
    dfs(root, false, 0);
    // dfs(root, true, 0); would also work
    return pathLength;
  }

  public void dfs(TreeNode node, boolean goLeft, int steps) {
    if (node == null)
      return;
    pathLength = Math.max(pathLength, steps);
    // node was right child
    if (goLeft) {
      dfs(node.left, false, ++steps);
      dfs(node.right, true, 1);
    }
    else {
      dfs(node.left, false, 1);
      dfs(node.right, true, ++steps);
    }
  }
}

File: Problems/1372. (M) Longest ZigZag Path in Binary Tree/Python3/solution.py
class Solution:
  def longestZigZag(self, root: Optional[TreeNode]) -> int:
    # recursive DFS
    pathLength = 0

    def dfs(node, goLeft, steps):
      if not node:
        return
      nonlocal pathLength
      pathLength = max(pathLength, steps)
      # node was right child
      if goLeft:
        dfs(node.left, False, steps+1)
        dfs(node.right, True, 1)
      # node was left child
      else:
        dfs(node.left, False, 1)
        dfs(node.right, True, steps+1)
    
    # root has no parent, so direction does not matter
    dfs(root, False, 0)
    # dfs(root, True, 0) would also work

    return pathLength

